### ISSUE TITLE
typechecking in the `geometry` column only accepts shapely shapes

### ISSUE THREAD

**Issue by ljwolf** [geopandas/geopandas#791]
Hey,

pysal/pysal#1054 calls this issue into clarity.

If you pack a dataframe with a column of non-`shapely` geometries that still have a [`__geo_interface__`](https://gist.github.com/sgillies/2217756), it'd be nice for `geopandas` to attempt to `shapely.geometry.asShape` on them before bailing out.

I can make that patch, but is that desirable?

There aren't too many non-shapely geometry packages left for spatial data representation, (I think `pygeoif` and `pysal` are the only two), but it seems like throwing an error that geometries aren't `valid` if they're perfectly valid OGC and expose `__geo_interface__` is surprising. 

Our polygon-in-dataframe is exposed in `pysal.lib.io.read_files`, and we're strongly discouraging its use for the future since we think geopandas should be the standard. But, in addition to handling this rough edge, it'd be useful for people who query a plaintext database with a `VARCHAR` column of `geometry` in json. 

For example
```python
import libpysal
import geopandas
non_shapely_df = libpysal.io.geotable.read_files(libpysal.examples.get_path("columbus.shp"))
test = geopandas.GeoDataFrame(non_shapely_df) # does not try to convert geometry
test = geopandas.GeoDataFrame(non_shapely_df, geomery='geometry') # fails on TypeError
```

**Comment by jwilson8767** [comment]
The `geometry` option when creating a GeoDataFrame can be any series, so if you're just concerned about solving this one instance I'm betting something like `test = geopandas.GeoDataFrame(non_shapely_df, geometry=non_shapely_df.apply(lambda g: shapely.geometry.asShape(g))`. 

That said, I agree that GeoDataFrames should handle initial type errors with a bit more forgiveness (though I'm not totally sure that means passing them to shapely for interpretation).

**Comment by jorisvandenbossche** [comment]
Additional / related problem is that in other places, we don't do any validation at all. Eg if you don't specify the column explicitly (but by default it still takes the one named 'geometry'):

```
In [12]: df = geopandas.GeoDataFrame({'a': [1, 2, 3], 'geometry': [(1, 1), (2, 2), (3, 3)]})

In [13]: df
Out[13]: 
   a geometry
0  1   (1, 1)
1  2   (2, 2)
2  3   (3, 3)
```

which 'works' but will later fail on other operations.

So I agree that upon construction, we should:

- always ensure that the final result actually contains valid geometry values (or a missing value indicator), thus coercing/check the input
- when coercing the input, we can not only accept shapely objects, but also convert anything that has the `__geo_interface__` (and only holds a single geometry) to shapely objects.

(note: in the cython branch, we are already more strict about this and the geometry column is always actually holding geometry objects, but there we also only check for shapely objects)

**Comment by ljwolf** [comment]
closed by #995
